Adding new cutlass configurations for llama70B

kushanam opened the pull request:
FILL IN THE PR DESCRIPTION HERE

A new set of cutlass kernel configurations (m,n,k) for best performance with llama 70B-x.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

terrynv committed:
new cutlass kernels for llama70B

kushanam committed:
fix clang formatting

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:
```diff
@@ -603,6 +660,172 @@ void cutlass_gemm_sm90_fp8_dispatch(torch::Tensor& out, torch::Tensor const& a,
+    }
+  }  // end of N = 8192, K = 8192
+
+  printf("Did not find a config for: M=%d, N=%d, K=%d\n", M, N, K);
```
reminder to remove debug cruft before landing

varun-sundar-rabindranath commented on csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:
```diff
@@ -603,6 +660,172 @@ void cutlass_gemm_sm90_fp8_dispatch(torch::Tensor& out, torch::Tensor const& a,
+      return cutlass_gemm_caller<Cutlass3xGemmP3>(
+          out, a, b, std::forward<EpilogueArgs>(args)...);
+    }
+  }  // end of N = 8192, K = 8192
```
Can you move this logic to a function in a .cuh file (something like https://github.com/vllm-project/vllm/blob/main/csrc/quantization/cutlass_w8a8/scaled_mm_c2x_sm80_dispatch.cuh) and then call the that function based on N and K values ? 

I believe it'll be better of maintainability and the logic in this function will be clearer.

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:
```diff
@@ -511,6 +511,63 @@ struct sm90_fp8_config_M64 {
+                      KernelSchedule, EpilogueSchedule>;
+};
+
+template <typename InType, typename OutType,
+          template <typename, typename, typename> typename Epilogue>
+struct sm90_fp8_config_P1 {
+  static_assert(std::is_same<InType, cutlass::float_e4m3_t>());
+  using KernelSchedule =
+      cutlass::gemm::KernelTmaWarpSpecializedPingpongFP8FastAccum;
+  using EpilogueSchedule = typename cutlass::epilogue::TmaWarpSpecialized;
+  using TileShape = Shape<_64, _128, _128>;
+  using ClusterShape = Shape<_2, _1, _1>;
+  using Cutlass3xGemm =
+      cutlass_3x_gemm<InType, OutType, Epilogue, TileShape, ClusterShape,
+                      KernelSchedule, EpilogueSchedule>;
+};
```
Is this the same as `sm90_fp8_config_M128`?

kushanam replied:
Yes, Removed the redundant config.

tlrmchlsmth commented on csrc/quantization/cutlass_w8a8/scaled_mm_c3x.cu:
```diff
@@ -511,6 +511,63 @@ struct sm90_fp8_config_M64 {
+                      KernelSchedule, EpilogueSchedule>;
+};
+
+template <typename InType, typename OutType,
+          template <typename, typename, typename> typename Epilogue>
+struct sm90_fp8_config_P2 {
+  static_assert(std::is_same<InType, cutlass::float_e4m3_t>());
+  using KernelSchedule =
+      cutlass::gemm::KernelTmaWarpSpecializedPingpongFP8FastAccum;
+  using EpilogueSchedule = typename cutlass::epilogue::TmaWarpSpecialized;
+  using TileShape = Shape<_128, _128, _128>;
+  using ClusterShape = Shape<_2, _1, _1>;
+  using Cutlass3xGemm =
+      cutlass_3x_gemm<InType, OutType, Epilogue, TileShape, ClusterShape,
+                      KernelSchedule, EpilogueSchedule>;
+};
```
Is this the same as `sm90_fp8_config_default`?

kushanam replied:
Yes, Removed the redundant config.

tlrmchlsmth reviewed (commented):
Hi @kushanam, thanks for the contribution! The configs look pretty good from the perspective of this being the right way to add configs and tune for specific cases.

Could you share e2e performance for both H100 and H200 for llama70b? Micro-benchmarks would be greatly appreciated as well. You could use the script `benchmarks/cutlass_benchmarks/w8a8_benchmarks.py` for this.

I'm not opposed to specializing the kernel configs specifically for Llama70B, but do you think it would be possible to make these cases more general? A couple of the new kernel definitions are duplicates of existing ones, so it'd be nice if they could be unified.

Lastly, is this only for the tensor parallelism == 1 case of Llama70B? Do you think these configs could be generalized to handle TP==2 and TP==4 as well?

kushanam committed:
removing redundant kernel configs and adding tests

kushanam commented:
> Hi @kushanam, thanks for the contribution! The configs look pretty good from the perspective of this being the right way to add configs and tune for specific cases.
> 
> Could you share e2e performance for both H100 and H200 for llama70b? Micro-benchmarks would be greatly appreciated as well. You could use the script `benchmarks/cutlass_benchmarks/w8a8_benchmarks.py` for this.
> 
> I'm not opposed to specializing the kernel configs specifically for Llama70B, but do you think it would be possible to make these cases more general? A couple of the new kernel definitions are duplicates of existing ones, so it'd be nice if they could be unified.
> 
> Lastly, is this only for the tensor parallelism == 1 case of Llama70B? Do you think these configs could be generalized to handle TP==2 and TP==4 as well?

For Microbenchmarks:

![image (5)](https://github.com/user-attachments/assets/ee0c8bd2-efdb-4049-ba30-a4e81202600c)

For e2e, llama70B : 1-2% faster on a single H200 instance

As for generalization and per our discussion (also addressing the comment by @varun-sundar-rabindranath, will revisit the whole structure and refactor in the next PR.

As per kernel benchmarks the main improvement for this PR comes from larger M sizes. Some of the optimizations would still apply but the improvement is less significant. I believe once a look-up-like structure in place, we can further optimize for finer grainer sizes
